Modules/arraymodule.c

```diff
@@ -2661,19 +2804,37 @@ array_ass_subscr(PyObject *op, PyObject *item, PyObject *value)
     }
 }
 
+static int
+array_ass_subscr(PyObject *op, PyObject* item, PyObject* value)
+{
+    int ret;
+    if (value == NULL) {
+        Py_BEGIN_CRITICAL_SECTION(op);
+        ret = array_ass_subscr_lock_held(op, item, value);
+        Py_END_CRITICAL_SECTION();
+    }
+    else {
+        Py_BEGIN_CRITICAL_SECTION2(op, value);
```

Comment:
We only needto acquire critical section when value is array, we can do that in lock held variant after that check, no need to do always.

Reply:
I specifically did it this way for the possibility of something already holding a lock on `value` and attempting a `bytearray` operation, deadlock. If locking two objects they should always be locked in a deterministic order which is what `Py_BEGIN_CRITICAL_SECTION2` does. Or am I overthinking?

Reply:
I think I need to revisit `bytearray_ass_subscript` for this same reason, will do so tomorrow.

Reply:
Checking this actually made me go down a rabbit hole. TL;DR: Haven't been able to force one, so feel free to smack me with the stupid stick if this is not the case, but maybe there is a possibility of deadlock even with a list subclass object? For reference, here is list assign subscript:

https://github.com/python/cpython/blob/a1d4b1c44741bae5bf93009223f3377aa45846b5/Objects/listobject.c#L3779

Given a list subclass:
```
class mylist(list):
    pass
```
And two `mylist` objects `a` and `b`, is a deadlock not possible if thread 1 does `a[:] = b` and thread 2 does `b[:] = a` simultaneously (since each will initially only take lock on self and later attempt lock on other through getitem)? Shouldn't the lock be taken on the second object always, just in case?

Reply:
Ping @colesbury , is what I wrote just above an issue for `list` or am I overthinking?

Reply:
No, it won't deadlock. See https://peps.python.org/pep-0703/#python-critical-sections and https://docs.python.org/3/c-api/init.html#python-critical-section-api

Nested critical sections pose atomicity concerns, but they won't cause deadlock. When a thread blocks trying to acquire a lock, it will release any active critical sections it holds. So for example:

```
Py_BEGIN_CRITICAL_SECTION(a);
a->foo += 3; // ok, holds the lock on a

  Py_BEGIN_CRITICAL_SECTION(b);  // may release the lock on a
  b->bar += 1; // ok, holds the lock on b
  // thread may not hold the lock on a here!
  Py_END_CRITICAL_SECTION(); // ensures that the lock on a is held again

a->foo -= 3; // ok, we hold the lock on a again after the Py_END_CRITICAL_SECTION()
Py_END_CRITICAL_SECTION();